src/clr-angular/forms/password/password.spec.ts

```diff
@@ -0,0 +1,96 @@
+import { ClrCommonFormsModule } from '../common/common.module';
+import { WrappedFormControl } from '../common/wrapped-control';
+import { ControlIdService } from '../common/providers/control-id.service';
+
+@Component({
+  template: `
+    <input type="password" clrPassword />
+  `,
+})
+class InvalidUseTest {}
+
+@Component({
+  template: `
+    <clr-password-container>
+      <input type="text" clrPassword name="model" [(ngModel)]="model" />
+    </clr-password-container>
+  `,
+})
+class SimpleTest {}
+
+export default function(): void {
+  describe('Password directive', () => {
+    let password, fixture, ifErrorService, ngControlService;
+
+    describe('invalid use', () => {
+      it('throws error when used without a form control', () => {
+        TestBed.configureTestingModule({ declarations: [ClrPassword, InvalidUseTest] });
+        expect(() => {
+          fixture = TestBed.createComponent(InvalidUseTest);
+          fixture.detectChanges();
+        }).toThrow();
+      });
+    });
+
+    describe('basic use', () => {
+      beforeEach(() => {
+        spyOn(WrappedFormControl.prototype, 'ngOnInit');
+        TestBed.configureTestingModule({
+          imports: [FormsModule, ClrIconModule, ClrCommonFormsModule],
+          declarations: [ClrPassword, ClrPasswordContainer, SimpleTest],
+          providers: [IfErrorService, NgControlService, ControlIdService],
+        });
+        fixture = TestBed.createComponent(SimpleTest);
+        password = fixture.debugElement.query(By.directive(ClrPassword));
+        ifErrorService = password.injector.get(IfErrorService, null);
+        ngControlService = password.injector.get(NgControlService, null);
+        spyOn(ifErrorService, 'triggerStatusChange');
+        spyOn(ngControlService, 'setControl');
+        fixture.detectChanges();
+      });
+
+      it('should apply the clr-password class', () => {
+        expect(password.nativeElement.classList.contains('clr-password'));
+      });
+
+      it('should have the IfErrorService', () => {
+        expect(ifErrorService).toBeTruthy();
+      });
+
+      it('should have the NgControlService and set the control', () => {
+        expect(ngControlService).toBeTruthy();
+        expect(ngControlService.setControl).toHaveBeenCalled();
+      });
+
+      it('correctly extends WrappedFormControl with ClrPasswordContainer', () => {
+        expect(password.injector.get(ClrPassword).wrapperType).toBe(ClrPasswordContainer);
+        expect(WrappedFormControl.prototype.ngOnInit).toHaveBeenCalled();
+      });
+
+      it('should handle blur events', () => {
+        // passwords must be both invalid and blurred to register the validity
+        password.nativeElement.value = 'abc';
+        password.nativeElement.dispatchEvent(new Event('input'));
+        password.nativeElement.dispatchEvent(new Event('blur'));
+        fixture.detectChanges();
+        expect(ifErrorService.triggerStatusChange).toHaveBeenCalled();
+      });
+    });
```

Comment:
Maybe add a test to check we properly set the `type="password"` attribute on initialization.

Reply:
Done